[Feature, Refactor]Order shopping list items by time frame + Code Organization

jendevelops opened the pull request:
## Description

- Abstract out shoppingListInput as it's own separate component called ShoppingListItem for better clarity and modularity
- Change ShoppingList html elements from unordered list to a table, so users can better utilize all the information stored in firebase in an easy to read format
- Abstracted the normalizeString function from AddItemForm into /lib/normalizeString.js as a helper function that is available app-wide
- Create helper function filterShoppingListByTimeframe that prefilters the shopping list (utilizes normalizeString previously added to /lib/) taken from firebase before setting to local shoppingListItems state

## Related Issue

Closes issue #13 

## Acceptance Criteria

- [x] Items in the list are shown visually grouped in the following order: Soon, Kind of soon, Not soon, Inactive (Still needs to consider inactive shopping list items)

- [x] Within each grouping, items should be sorted by the estimated number of days until next purchase

- [x] Items with the same number of estimated days until next purchase should be sorted alphabetically

- [x] Items in the different states should be visually distinct

- [x] Items in the different states should be described distinctly when read by a screen reader


## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓      | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

jendevelops committed:
Initialize ShoppingListItem component

jendevelops committed:
Move shoppingListItemInput into its own component ShoppingListItem

jendevelops committed:
Change shopping list item display format from ul to table

jendevelops committed:
Add timeFrameString for timeframe text in table

jendevelops committed:
Add filter logic helper function, and filter shoppingList before setting local state

dkimlim committed:
added sorting logic for alphabetical itemNames based on timeframe

dkimlim committed:
Abstract normalizeString and Sort itemName alphabetically

Create helper function in lib called normalizeString,removing it
from AddItemForm. Now function is available app wide. Then, it was
utilized in alphabeticalSort function to filter shoppinglist items
by name.

laurenmbeatty commented on src/pages/ShoppingList.js:
```diff
@@ -57,6 +43,23 @@ const ShoppingList = ({ token }) => {
+        //does not consider a case where timeFrame is none of these
+
+        return seven.concat(fourteen).concat(thirty);
+    };
```
V. cool!!!!  I think you could also do `return [...seven, ...fourteen, ...thirty]`. Just an alternative, though.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

dkimlim committed:
Add aria-label to checkbox input in ShoppingListItems component

created aria-label for accessibility screenreaders that will read
a string with the item's name and timeframe.

dkimlim committed:
Enabled INACTIVE state if last purchase was 2 x the estimated time

Updated the local shopping item list to remove the inactive items
before filtering them by timeframe. Added css for each timeframe
in the UI of shopping list.

laurenmbeatty commented on src/components/ShoppingListItem.js:
```diff
@@ -0,0 +1,68 @@
+            return "None";
+    }
+}
+// const getStringValue = (numberValue) => {
```
Maybe can rid of this? Leave it if you think you will need it, but otherwise you can remove.

laurenmbeatty commented on src/components/AddItemForm.js:
```diff
@@ -3,6 +3,7 @@ import { Redirect } from 'react-router-dom';
 import '../css/AddItemForm.css';
 import moment from 'moment';
+import normalizeString from '../lib/normalizeString';
 
```
Yes!!!!!  Love it.

laurenmbeatty commented on src/components/ShoppingListItem.js:
```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+import * as timeframeConstants from '../lib/timeframeConstants';
+
```
Are you using `timeFrameConstants` anymore?

jendevelops replied:
We plan on utilizing the constants throughout the app once fully implemented.

laurenmbeatty replied:
Sounds good. If you're not using it in this file right now, though, you can remove line 2.

laurenmbeatty commented on src/components/ShoppingListItem.js:
```diff
@@ -0,0 +1,68 @@
+
+
+const ShoppingListItem = ({item,handleCheck}) => {
+    const ariaString = `${item.itemName} to be bought in ${timeFrameString(item.timeFrame)}`;
```
Nice. Love that you separated this out into its own file.

laurenmbeatty commented on src/components/ShoppingListItem.js:
```diff
@@ -0,0 +1,68 @@
+import * as timeframeConstants from '../lib/timeframeConstants';
+
+const timeFrameString = number =>
+{
```
This is not a blocker at all, but maybe these 2 functions could be combined into one since they both do essentially the same thing? The case could return an object that has both the css class and the label in it? idk.

jendevelops replied:
Yes, I was thinking the same thing. I was thinking of dropping the values in 2 dictionaries with the numbers as the keys, and having a 2nd argument to determine what you want returned.

laurenmbeatty commented:
Great commit messages!!!!!!!

AshlingH commented on src/lib/timeframeConstants.js:
```diff
@@ -0,0 +1,6 @@
+const KIND_OF_SOON = {stringValue: "between 7 and 30 days", integerValue: 14}
+const NOT_SOON = {stringValue: "more than 30 days", integerValue: 30}
+const INACTIVE = {stringValue: "inactive", integerValue: 0}
+
```
Brings a full circle feel to an earlier issue for Diane and I, thank you!

laurenmbeatty replied:
Awesome. Great idea.

AshlingH commented on src/css/ShoppingList.css:
```diff
@@ -0,0 +1,12 @@
+}
+.inactive {
+    background-color: grey;
+}
```
From a user perspective, this is soooo helpful well done!

AshlingH reviewed (approved):
Really good jobs guys, looking forward to hearing about your process on Sunday. As the app develops issues and items from earlier in the cohort are coming together for me. Nice approach!

jonbascos reviewed (commented):
Excellent work you two.  Love the new filterShoppingListByTimeframe function.  Looks pretty clean.  Can't wait to hear about it from you on Sunday.

laurenmbeatty commented on src/components/ShoppingListItem.js:
```diff
@@ -0,0 +1,68 @@
+        <tr className={generateClass(item.timeFrame)}>
+            <td>
+                <input
+                aria-label= {ariaString}
```
Great!

laurenmbeatty commented on src/pages/ShoppingList.js:
```diff
@@ -57,6 +44,32 @@ const ShoppingList = ({ token }) => {
+        const fourteen = shoppingListArray.filter(item => item.timeFrame === 14).sort(alphabeticalSort);
+        const thirty = shoppingListArray.filter(item => item.timeFrame === 30).sort(alphabeticalSort);
+        const inactive = shoppingListArray.filter(item => item.timeFrame === 0).sort(alphabeticalSort);
+
```
Love these one-liners. Very nice.

laurenmbeatty commented:
Screen reader info is perfect. nice work.

laurenmbeatty commented:
![image](https://user-images.githubusercontent.com/10378331/79625609-b4f95680-80ef-11ea-8740-9006dfcc5ee5.png)

laurenmbeatty reviewed (approved):
This looks amazing. 💖

dkimlim committed:
Merge branch 'master' of https://github.com/the-collab-lab/tcl-5-smart-shopping-list into jvb-dk-sort-by-timeframe
